components/tikv_util/src/sys/thread.rs

```diff
@@ -181,7 +181,120 @@ mod imp {
     }
 }
 
-#[cfg(not(target_os = "linux"))]
+#[cfg(target_os = "macos")]
+#[allow(bad_style)]
+mod imp {
+    use std::io;
+    use std::iter::FromIterator;
+    use std::mem::size_of;
+    use std::ptr::null_mut;
+    use std::slice;
+
+    use libc::*;
+
+    pub type Pid = mach_port_t;
+
+    type task_inspect_t = mach_port_t;
+    type thread_act_t = mach_port_t;
+    type thread_act_array_t = *mut thread_act_t;
+
+    extern "C" {
+        fn task_threads(
+            target_task: task_inspect_t,
+            act_list: *mut thread_act_array_t,
+            act_listCnt: *mut mach_msg_type_number_t,
+        ) -> kern_return_t;
+    }
+
+    #[derive(Default)]
+    pub struct FullStat {
+        pub stime: i64,
+        pub utime: i64,
+        pub command: String,
+    }
+
+    #[inline]
+    pub fn clock_tick() -> i64 {
+        1_000_000
+    }
+
+    /// Gets the ID of the current process.
+    #[inline]
+    pub fn process_id() -> Pid {
+        unsafe { mach_task_self_ }
+    }
+
+    /// Gets the ID of the current thread.
+    #[inline]
+    pub fn thread_id() -> Pid {
+        unsafe { mach_thread_self() }
+    }
+
+    pub fn thread_ids<C: FromIterator<Pid>>(pid: Pid) -> io::Result<C> {
+        // https://www.gnu.org/software/hurd/gnumach-doc/Task-Information.html
+
+        unsafe {
+            let mut act_list: thread_act_array_t = null_mut();
+            let mut act_count: mach_msg_type_number_t = 0;
+            let ret = task_threads(pid, &mut act_list, &mut act_count);
+            if ret != KERN_SUCCESS {
+                return Err(io::Error::from_raw_os_error(ret));
+            }
+
+            let pids = slice::from_raw_parts_mut(act_list, act_count as _)
+                .iter()
+                .copied()
+                .collect();
+
+            vm_deallocate(
+                pid,
+                act_list as vm_address_t,
+                size_of::<thread_act_t>() * act_count as usize,
+            );
+
+            Ok(pids)
+        }
+    }
+
+    pub fn full_thread_stat(_pid: Pid, tid: Pid) -> io::Result<FullStat> {
+        // https://www.gnu.org/software/hurd/gnumach-doc/Thread-Information.html
+
+        unsafe {
+            let flavor = THREAD_BASIC_INFO;
+            let mut info = std::mem::zeroed::<thread_basic_info>();
+            let mut thread_info_cnt =
+                (size_of::<thread_basic_info>() / size_of::<natural_t>()) as mach_msg_type_number_t;
```

Comment:
You mean `THREAD_BASIC_INFO_COUNT`?

Reply:
yes, and not found from libc

Reply:
Interesting, but it does compile when using the constant.

Reply:
updated